### ISSUE TITLE
A node type was not returned. This is either because a gradient was not defined, or the input is independent of the output

### ISSUE THREAD

**Issue by jarmstrong2** [twitter/torch-autograd#74]
I am basically mimicking your example code for autocriterion, but with my own function, but i get an error, here is the function in question:
```
autograd = require 'autograd'

local cmd = torch.CmdLine()

cmd:text()
cmd:text('Script for training model.')

cmd:option('-inputSize' , 61, 'number of input dimension')
cmd:option('-dimSize' , 2, 'dim size for U')
cmd:option('-batchSize' , 4, 'mini batch size')
cmd:option('-numMixture' , 10, 'number of mixture components in output layer')

cmd:text()
opt = cmd:parse(arg)

function logDet(matrix)
	output = torch.zeros(opt.batchSize, 1)
    for i = 1, opt.batchSize do
        eig_vals = (torch.eig(matrix[i], 'N'))
        output[i] = (torch.log(eig_vals:select(2, 1))):sum()
    end
    return output
end

function inverse(matrix)
    output = torch.zeros(opt.batchSize, opt.inputSize, opt.inputSize)
    for i = 1, opt.batchSize do
        output[i] = (torch.inverse(matrix[i]))
    end
    return output
end

function getEps()
   eps = torch.eye(opt.inputSize,opt.inputSize) * 1e-2
   eps:resize(1,opt.inputSize,opt.inputSize)
   fulleps = eps:clone()
   for i = 2, opt.batchSize do
       fulleps = torch.cat(fulleps,eps,1)
    end
    return fulleps
end

local mixtureMultvarGauss = function(input, target)
	local sizeMeanInput = opt.inputSize * opt.numMixture
	local sizeCovarianceInput = opt.inputSize * opt.numMixture * opt.dimSize

	local piStart = 1
	local piEnd = opt.numMixture
	local hat_pi_t = input[{{},{piStart,piEnd}}]

	local muStart = piEnd + 1
	local muEnd = piEnd + sizeMeanInput
	local hat_mu_t = input[{{},{muStart,muEnd}}]

	local sigmaStart = muEnd + 1
	local sigmaEnd = muEnd + sizeCovarianceInput
	local hat_sigma_t = input[{{},{sigmaStart,sigmaEnd}}]

	local mask = input[{{},{sigmaEnd + 1}}]

	hat_mu_t:resize(opt.batchSize, opt.numMixture, 1, opt.inputSize)
	hat_sigma_t:resize(opt.batchSize, opt.numMixture, opt.dimSize, opt.inputSize)
	target:resize(opt.batchSize, 1, opt.inputSize)
	eps = getEps()

	local join_mixture_result = torch.zeros(opt.batchSize, opt.numMixture)

	for i = 1, opt.numMixture do
		local u = hat_sigma_t[{{},{i},{},{}}]:squeeze(2)
		local mu = hat_mu_t[{{},{i},{},{}}]:squeeze(2)
		local pi = hat_pi_t[{{},{i}}]
		local sigma = torch.bmm(u:transpose(2,3), u)
		sigma:add(eps)
		local det_sigma_2_pi = logDet(sigma) + (opt.inputSize * torch.log(2 * math.pi)) 
		local sqr_det_sigma_2_pi = (det_sigma_2_pi) * -0.5

		local target_mu = target - mu 
		local transpose_target_mu = target_mu:transpose(2,3)

		local inv_sigma = inverse(sigma)
		local transpose_target_mu_sigma = torch.bmm(target_mu, inv_sigma)

		local transpose_target_mu_sigma_target_mu = torch.bmm(transpose_target_mu_sigma, 
			transpose_target_mu)

		local exp_term = transpose_target_mu_sigma_target_mu * -0.5

		local mixture_result = pi + sqr_det_sigma_2_pi + exp_term

        join_mixture_result[{{},{i}}] = mixture_result
	end

	local max_mixture = torch.max(join_mixture_result, 2)
	local max_expanded = max_mixture:expandAs(join_mixture_result) * -1
	local norm_mixture = max_expanded + join_mixture_result
	local norm_mixture_exp = torch.exp(norm_mixture)
	local norm_mixture_sumexp = torch.sum(norm_mixture_exp, 2)
	local norm_mixture_logsumexp = torch.log(norm_mixture_sumexp)
	local norm_mixture_addlogsumexp = (max_mixture + norm_mixture_logsumexp)*-1

	return torch.cmul(mask, norm_mixture_addlogsumexp)
end 

local autoMseCriterion = autograd.nn.AutoCriterion('AutoMixGauss')(mixtureMultvarGauss)

pi = torch.rand(4,10)
mask = torch.ones(4,1)
sig = torch.rand(4, 61*2*10)
mu = torch.rand(4, 61*10)
input = torch.cat(pi, mu, 2)
input = torch.cat(input, sig, 2)
input = torch.cat(input, mask, 2)
target = torch.rand(4,61)
print(autoMseCriterion:forward(input, target))
```

**Comment by alexbw** [comment]
Assignment via __newindex is not allowed in autograd functions:

So:

```lua
 join_mixture_result[{{},{i}}] = mixture_result
```

is illegal. We're thinking on ways of providing better error messages for this.

**Comment by jarmstrong2** [comment]
What would you recommend as an alternative to assigning values to an already defined tensor? I did try torch.cat as well but without success.

**Comment by fmassa** [comment]
@jarmstrong2 maybe you can use `select`/`narrow` + `copy` ?

**Comment by alexbw** [comment]
Why did torch.cat fail?

**Comment by jarmstrong2** [comment]
it gave the same error, so it's a bit of a mystery to me also.
